drivers/ethernet/dsa_ksz8794.c

```diff
@@ -0,0 +1,579 @@
+
+	if (!swspi->is_init) {
+		return -ENODEV;
+	}
+
+	dsa_ksz8794_get_static_mac_table(swspi, buf, tbl_entry_idx);
+
+	return 0;
+}
+
+#if defined(CONFIG_DSA_KSZ8794_TAIL_TAGGING)
+#define DSA_KSZ8795_TAIL_TAG_OVERRIDE	BIT(6)
+#define DSA_KSZ8795_TAIL_TAG_LOOKUP	BIT(7)
+
+#define DSA_KSZ8794_EGRESS_TAG_LEN 1
+#define DSA_KSZ8794_INGRESS_TAG_LEN 1
+
+#define DSA_MIN_L2_FRAME_SIZE 64
+#define DSA_L2_FCS_SIZE 4
+
+struct net_pkt *dsa_xmit_pkt(struct net_if *iface, struct net_pkt *pkt)
+{
+	struct ethernet_context *ctx = net_if_l2_data(iface);
+	struct net_eth_hdr *hdr = NET_ETH_HDR(pkt);
+	struct net_linkaddr lladst;
+	struct net_buf *buf;
+	uint8_t port_idx, i;
+	size_t len, pad;
+
+	lladst.len = sizeof(hdr->dst.addr);
+	lladst.addr = &hdr->dst.addr[0];
+
+	len = net_pkt_get_len(pkt);
+	/*
+	 * For KSZ8794 one needs to 'pad' the L2 frame to its minimal size
+	 * (64B) before appending TAIL TAG and FCS
+	 */
+	if (len < (DSA_MIN_L2_FRAME_SIZE - DSA_L2_FCS_SIZE)) {
+		/* Calculate number of bytes needed for padding */
+		pad = DSA_MIN_L2_FRAME_SIZE - DSA_L2_FCS_SIZE - len;
+		buf = net_buf_alloc_len(net_buf_pool_get(pkt->buffer->pool_id),
+					pad, K_NO_WAIT);
+
+		if (!buf) {
+			LOG_ERR("DSA cannot allocate new data buffer");
+			return NULL;
+		}
+
+		buf = net_buf_frag_last(net_buf_frag_add(pkt->buffer, buf));
+		for (i = 0; i < pad; i++)
+			net_buf_add_u8(buf, 0x0);
+	}
+
+	buf = net_buf_alloc_len(net_buf_pool_get(pkt->buffer->pool_id),
+				DSA_KSZ8794_INGRESS_TAG_LEN, K_NO_WAIT);
+	buf = net_buf_frag_last(net_buf_frag_add(pkt->buffer, buf));
+
+	/*
+	 * For master port (eth0) set the bit 7 to use look-up table to pass
+	 * packet to correct interface (bits [0..6] _are_ ignored).
+	 *
+	 * For slave ports (lan1..3) just set the tag properly:
+	 * bit 0 -> eth1, bit 1 -> eth2. bit 2 -> eth3
+	 * It may be also necessary to set bit 6 to "anyhow send packets to
+	 * specified port in Bits[3:0]". This may be needed for RSTP
+	 * implementation (when the switch port is disabled, but shall handle
+	 * LLDP packets).
+	 */
+	if (DSA_IS_PORT_MASTER(iface)) {
+		port_idx = DSA_KSZ8795_TAIL_TAG_LOOKUP;
+	} else {
+		port_idx = (1 << (ctx->dsa_port_idx - 1));
+	}
+
+	NET_DBG("TT - port: 0x%x[%p] LEN: %d 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x",
+		port_idx, iface, len, lladst.addr[0], lladst.addr[1],
+		lladst.addr[2], lladst.addr[3], lladst.addr[4], lladst.addr[5]);
+
+	net_buf_add_u8(buf, port_idx);
+	return pkt;
+}
+
+struct net_if *dsa_get_iface(struct net_if *iface, struct net_pkt *pkt)
+{
+	struct ethernet_context *ctx;
+	struct net_if *iface_sw;
+	size_t plen;
+	uint8_t pnum;
+
+	if (!(net_eth_get_hw_capabilities(iface) &
+	      (ETHERNET_DSA_SLAVE_PORT | ETHERNET_DSA_MASTER_PORT))) {
+		return iface;
+	}
+
+	net_pkt_cursor_init(pkt);
+	plen = net_pkt_get_len(pkt);
+
+	net_pkt_skip(pkt, plen - DSA_KSZ8794_EGRESS_TAG_LEN);
+	net_pkt_read_u8(pkt, &pnum);
+
+	net_pkt_update_length(pkt, plen - DSA_KSZ8794_EGRESS_TAG_LEN);
+	plen = net_pkt_get_len(pkt);
```

Comment:
avoid calling net_pkt_get_len again here, just update plen accordingly. net_pkt_get_len is always going through the net_buf linked list and is thus not optimized

Reply:
I could remote this call from the line 517, as it is used only in the NET_DBG() macro for debugging.
It can be replaced wtih plen - DSA_KSZ8794_EGRESS_TAG_LEN in the NET_DBG() call.